File: Difficulty: Medium/Strongly Connected/strongly-connected.cpp
//Position this line where user code will be pasted.
class Solution {
    private:
    void dfs(vector<vector<int>> &adj, vector<bool> &visited,stack<int> &st, int src){
        visited[src]=true;
        
        for(auto &nbr: adj[src]){
            if(visited[nbr]==false){
                dfs(adj,visited,st,nbr);  
            } 
        }
        
        st.push(src);
    }
    
    void dfs2(vector<vector<int>> &adj, vector<bool> &visited, int src){
        visited[src]=true;
        
        for(auto &nbr: adj[src]){
            if(visited[nbr]==false){
                dfs2(adj,visited,nbr);  
            } 
        }
        
    }
    
  public:
    int kosaraju(vector<vector<int>> &adj) {
        // code here
        
        //I have to get the edges based on finish time 
        int V = adj.size();
        vector<bool> visited(V,false);
        stack<int> st;
        
        for(int i=0;i<V;i++){
            if(visited[i]==false){
                dfs(adj,visited,st,i);
            }    
        }
        
        
        //step one is to reverse all the edges 
        vector<vector<int>> adjList(V);
        
        for(int i=0;i<V;i++){
            for(int j=0;j<adj[i].size();j++){
                int v = adj[i][j];
                adjList[v].push_back(i);
            }
        }
        
        visited.assign(V,false);
        int scc=0;
        
        while(!st.empty()){
            int t = st.top();
            st.pop();
            if(visited[t]==false){
                dfs2(adjList,visited,t);
                scc++;
            }
        }
        
        return scc;
    }
};

File: Difficulty: Medium/Strongly Connected/strongly-connected.py
from collections import defaultdict

class Solution:
    
    def finishingTime(self, src, adj, visited, st):
        
        visited[src] = True
        
        # visit all nbrs if not visited
        for nbr in adj[src]:
            if not visited[nbr]:
                self.finishingTime(nbr, adj, visited, st)
        
        st.append(src)
        return st
        
    def dfs(self, src, visited, adj):
        visited[src] = True
        for nbr in adj[src]:
            if not visited[nbr]:
                self.dfs(nbr, visited, adj)

    def kosaraju(self, V, edges):
        # code here
        
        if V == 0:
            return 0
        if not edges:
            return V
            
        # form the adj list
        adj = defaultdict(list)
        reverse_adj = defaultdict(list)
        
        for edge in edges:
            u = edge[0]
            v = edge[1]
            
            adj[u].append(v)
            reverse_adj[v].append(u)
        
        
        # step 1 : sort all the vertices according to finishing time
        visited = [False] * V
        st = []
        for i in range(V):
            if not visited[i]:
                self.finishingTime(i, adj, visited, st)
                
        
        # now start the dfs again in the reversed graph and count the SCC
        visited = [False] * V 
        scc = 0
        while st:
            src = st.pop()
            if not visited[src]:
                self.dfs(src, visited, reverse_adj)
                scc +=1
                
                
        return scc
            